wdl/transforms/biscayne/src/main/scala/wdl/transforms/biscayne/linking/expression/types/BiscayneTypeEvaluators.scala

```diff
@@ -161,14 +198,101 @@ object BiscayneTypeEvaluators {
-      // This requires some extra work to be done in a subsequent PR.
-      WomObjectType.validNel
+      evaluatedType match {
+        case Some(evaluated) =>
+          expectedType match {
+            case Some(expected) =>
+              if (areTypesAssignable(evaluated, expected)) evaluated.validNel
+              else
+                s"$typeName.$memberName expected to be ${expected.friendlyName}. Found ${evaluated.friendlyName}.".invalidNel
+            case None => s"Type $typeName does not have a member called $memberName.".invalidNel
+          }
+        case None => s"Error evaluating the type of ${typeName}.${memberName}.".invalidNel
+      }
+
+    // For each member in the literal, check that it exists in the struct definition and is the expected type.
+    def checkMembersAgainstDefinition(a: StructLiteral,
+                                      structDefinition: WomCompositeType,
+                                      linkedValues: Map[UnlinkedConsumedValueHook, GeneratedValueHandle],
+                                      typeAliases: Map[String, WomType]
+    ): ErrorOr[WomCompositeType] = {
+      val checkedMembers: Map[String, ErrorOr[WomType]] = a.elements.map { case (memberKey, memberExpressionElement) =>
+        val evaluatedType =
+          expressionTypeEvaluator.evaluateType(memberExpressionElement, linkedValues, typeAliases).toOption
+        val expectedType = structDefinition.typeMap.get(memberKey)
+        (memberKey, checkIfMemberIsValid(a.structTypeName, memberKey, evaluatedType, expectedType))
+      }
+
+      val (errors, validatedTypes) = checkedMembers.partition { case (_, errorOr) =>
+        errorOr match {
+          case Invalid(_) => true
+          case Valid(_) => false
+        }
+      }
+
+      errors.headOption match {
+        case Some((_, Invalid(_))) =>
+          errors.collect { case (_, Invalid(e)) => e.toList.mkString }.toList.mkString(",").invalidNel
+        case _ =>
+          val types = validatedTypes.collect { case (key, Valid(v)) => (key, v) }
+          WomCompositeType(types, Some(a.structTypeName)).validNel
+      }
+    }
+
+    // For every member in the definition, if that member isn't optional, confirm that it is also in the struct literal.
+    def checkForMissingMembers(foundMembers: Map[String, WomType],
+                               structDefinition: WomCompositeType
+    ): Option[String] = {
+      val errors: Iterable[String] = structDefinition.typeMap flatMap { case (memberName, memberType) =>
+        memberType match {
+          case WomOptionalType(_) => None
+          case _ =>
+            if (!foundMembers.contains(memberName)) Some(s"Expected member ${memberName} not found. ")
+            else None
+        }
+      }
+      errors match {
+        case Nil => None
+        case _ => Some(errors.mkString)
+      }
+    }
+    override def evaluateType(a: StructLiteral,
```

Comment:
I think this could use a comment describing the type validation strategy used, to help readers understand the roles of the methods above.